modules/sasl.cpp

```diff
@@ -338,10 +463,256 @@ class CSASLMod : public CModule {
+    PKCS5_PBKDF2_HMAC(sPassword.c_str(), sPassword.length(),
+                      (unsigned char*)sSalt.c_str(), ulSaltLen, uiIterCount,
+                      m_pDigest, m_uDigestSize, m_pSaltedPassword);
+
+    // AuthMessage := client-first-message-bare + "," +
+    //                server-first-message + "," +
+    //                client-final-message-without-proof
+    sClientFinalMessageWithoutProof = "c=biws,r=" + sServerNonceB64;
+
+    m_sAuthMessage = m_sClientFirstMessageBare + "," + sInput + "," +
+                     sClientFinalMessageWithoutProof;
+
+    // ClientKey := HMAC(SaltedPassword, "Client Key")
+    pClientKey = static_cast<unsigned char*>(malloc(m_uDigestSize));
+    HMAC(m_pDigest, m_pSaltedPassword, m_uDigestSize,
+         (unsigned char*)CLIENT_KEY, strlen(CLIENT_KEY), pClientKey,
+         &uiClientKeyLen);
+
+    // StoredKey := H(ClientKey)
+    if (!CreateSHA(pClientKey, m_uDigestSize, sStoredKey, &uiStoredKeyLen)) {
+        free(pClientKey);
+        return SCRAM_ERROR;
+    }
+
+    // ClientSignature := HMAC(StoredKey, AuthMessage)
+    pClientSignature = static_cast<unsigned char*>(malloc(m_uDigestSize));
+    memset(pClientSignature, 0, m_uDigestSize);
+    HMAC(m_pDigest, sStoredKey, uiStoredKeyLen,
+         (unsigned char*)m_sAuthMessage.c_str(), m_sAuthMessage.length(),
+         pClientSignature, NULL);
+
+    // ClientProof := ClientKey XOR ClientSignature
+    pClientProof = static_cast<unsigned char*>(malloc(uiClientKeyLen));
+    memset(pClientProof, 0, uiClientKeyLen);
+    for (uiIndex = 0; uiIndex < uiClientKeyLen; uiIndex++) {
+        pClientProof[uiIndex] = pClientKey[uiIndex] ^ pClientSignature[uiIndex];
+    }
+    sClientProofB64 = CString((const char*)pClientProof, uiClientKeyLen);
+    sClientProofB64.Base64Encode();
+
+    sOutput = sClientFinalMessageWithoutProof + ",p=" + sClientProofB64;
+    sOutput.Base64Encode();
+    m_pModule->SendAuthentication(sOutput);
+
+    free(pClientKey);
+    free(pClientSignature);
+    free(pClientProof);
+    m_uiStep++;
+    return SCRAM_IN_PROGRESS;
+}
+
+Scram::ScramStatus Scram::ProcessServerFinal(CString& sInput) {
+    CString sVerifier;
+    unsigned char *pServerKey, *pServerSignature;
+    unsigned int uiServerKeyLen = 0, uiServerSignatureLen = 0;
+    unsigned long ulVerifierLen;
+
+    if (sInput.length() < 3 || (sInput[0] != 'v' && sInput[1] != '=')) {
+        return SCRAM_ERROR;
+    }
+
+    sVerifier = sInput.substr(2);
+    ulVerifierLen = sVerifier.Base64Decode();
+
+    // ServerKey := HMAC(SaltedPassword, "Server Key")
+    pServerKey = static_cast<unsigned char*>(malloc(m_uDigestSize));
+    HMAC(m_pDigest, m_pSaltedPassword, m_uDigestSize,
+         (unsigned char*)SERVER_KEY, strlen(SERVER_KEY), pServerKey,
+         &uiServerKeyLen);
+
+    // ServerSignature := HMAC(ServerKey, AuthMessage)
+    pServerSignature = static_cast<unsigned char*>(malloc(m_uDigestSize));
```

Comment:
Please don't use memory manually like this, it's not C. Use std::string/std::vector/std::unique_ptr, etc